### PR TITLE
Migrate Scala CLI's using directive parser to a compiler module

### PR DESCRIPTION
**Pull request by Gedochao**
This moves the Scala CLI's directive parser (https://github.com/VirtusLab/scala-cli/tree/main/modules/directives-parser) to a module in the compiler repository.
It also migrates the REPL off the old parser to the new one, as well as gets rid of the  custom directive parser  (which wasn't at all fully compatible with how Scala CLI parses them) used by Vulpix (the compiler's testing harness).

## Have you relied on LLM-based tools in this contribution?
Yes, and I checked the output by running the whole test suite migrated from Scala CLI (plus some more)

## How was the solution tested?
New automated tests

### CODE REVIEW

**Comment by SolalPirelli** [comment]
Since we're officially supporting this it'd be good to centralize what's currently into comments into a more formal specification for what it accepts and doesn't accept? A readme or a page in the docs folder maybe?

File: directives-parser/README.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+original file where code begins.
+
+### Allowed before code
+
+- Blank lines
+- A `#!` shebang on the very first line only
+- Line comments (`// ...`, but not `//> using ...`)
+- Block comments (`/* ... */`), including multi-line and nested comments
+- Leading whitespace (spaces or tabs) before `//> using`
+- A UTF-8 BOM (`\uFEFF`) at the start of the file (stripped before parsing)
```
Comment (SolalPirelli) [inline directives-parser/README.md:39]: Do we support other BOMs, e.g., if the file is UTF-16? (I'm realizing while working on IO-related stuff that our `-encoding` support isn't the greatest)

Comment (Gedochao) [reply directives-parser/README.md:39]: https://github.com/VirtusLab/scala-cli/issues/1390 neither is Scala CLI's I guess 🤡

Comment (Gedochao) [reply directives-parser/README.md:39]: nope, we do not support other BOMs. legit feature request/follow-up, though. 
out of scope here IMO.

**Comment by Gedochao** [comment]
Note: I'm not including any sort of git history for the parser, because there is no history to speak of (it hasn't really been modified since I rewrote it from Java).

File: directives-parser/src/main/scala/dotty/tools/directives/UsingDirectivesParser.scala [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+object UsingDirectivesParser:
+
+  /** Parse a source file and return all using directives along with diagnostics.
+    *
+    * @param content
+    *   the raw content of the source file as a character array
+    * @return
+    *   a [[UsingDirectivesResult]] with parsed directives, code offset, and diagnostics
+    */
+  def parse(content: Array[Char]): UsingDirectivesResult =
```
Comment (SolalPirelli) [inline directives-parser/src/main/scala/dotty/tools/directives/UsingDirectivesParser.scala]: apologies for potentially being annoying but could this be, e.g., an `IndexedSeq[Char]`?

while simplifying dotty's IO I've noticed we often make our lives more complex by requiring arrays in APIs that don't need them, because, e.g., if you have a String you have no choice but to copy the entire String just to pass an array, even though the data is right there. And if you have a file then you have to read the entire file just in case, even if you end up only needing the first few bytes...

Comment (Gedochao) [reply directives-parser/src/main/scala/dotty/tools/directives/UsingDirectivesParser.scala]: Done.
Although I'd rather do theoretical further refactorings as follow-ups/separate PRs. 😅

Comment (SolalPirelli) [reply directives-parser/src/main/scala/dotty/tools/directives/UsingDirectivesParser.scala]: Thanks! Yeah refactorings can wait it's just that this one is the public API of the package ^^

File: project/Build.scala [hunk 3]
```diff
@@ -1355,13 +1357,62 @@ object Build {
...
+      autoScalaLibrary := false,
+      Compile / unmanagedSourceDirectories := Seq(baseDirectory.value / "src" / "main" / "scala"),
+      Test    / unmanagedSourceDirectories := Seq(baseDirectory.value / "src" / "test" / "scala"),
+      libraryDependencies ++= Seq(
+        Dependencies.sbtJunitInterface % Test,
+      ),
+      publish / skip := true,
+      target := target.value / "scala3-directives-parser-nonbootstrapped",
+      fetchedScalaInstanceSettings,
+      bspEnabled := false,
```
Comment (SolalPirelli) [inline project/Build.scala]: any reason not to enable it? I'd expect the nonbootstrapped to be always enabled and the bootstrapped to be conditional on the enableBspAllProjects or whatever it's called

File: directives-parser/README.md [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+- Directives inside block comments, line comments, or ScalaDoc comments
+- Directives after the first line of code (ignored with a warning)
+- A `package` declaration or any other code before directives (treats the file as
+  having no directive region; post-code directives warn)
+- `//> using` on the same line after a block comment closes (`/* comment */ //> using ...`)
+- Inline block comments within a directive line (not supported; `/*` becomes part of the key)
+
+### Prefix requirement
+
+The exact prefix `//> using` is required (with a single space after `>`):
```
Comment (SolalPirelli) [inline directives-parser/README.md:52]: is there some historical reason why we don't tolerate zero or more than one space?

File: compiler/test/dotty/tools/utils.scala [hunk 5]
```diff
@@ -138,39 +141,42 @@ private val toolArg = raw"(?://|/\*| \*) ?(?i:(${ToolName.values.mkString("|")})
 // =================================== VULPIX DIRECTIVES ==========================================
 // ================================================================================================
 
-/** Directive to specify to vulpix the options to pass to Dotty */
-private val directiveOptionsArg = raw"//> using option(?:s)? (.*)".r
-private val directiveJavacOptions = raw"//> using javacOpt (.*)".r
-private val directiveTargetOptions = raw"//> using target.platform (jvm|scala-js)".r
-private val directiveUnsupported = raw"//> using (scala) (.*)".r
-private val directiveUnknown = raw"//> using (.*)".r
```
Comment (SolalPirelli) [inline compiler/test/dotty/tools/utils.scala:146]: 🎉

File: directives-parser/src/main/scala/dotty/tools/directives/Diagnostics.scala [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package dotty.tools.directives
+
+enum DiagnosticSeverity:
+  case Error, Warning
+
+case class UsingDirectiveDiagnostic(
+  message: String,
+  severity: DiagnosticSeverity,
+  position: Option[Position] = None
```
Comment (SolalPirelli) [inline directives-parser/src/main/scala/dotty/tools/directives/Diagnostics.scala:9]: did I miss something or do all our diagnostics have a position?

File: directives-parser/src/main/scala/dotty/tools/directives/DirectiveValue.scala [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  case StringVal(value: String, isQuoted: Boolean, position: Position)
+
+  /** A boolean literal (`true` or `false`). */
+  case BoolVal(value: Boolean, position: Position)
+
+  /** No value was provided after the key (treated as `true` downstream). */
+  case EmptyVal(position: Position)
+
+object DirectiveValue:
+  extension (dv: DirectiveValue)
```
Comment (SolalPirelli) [inline directives-parser/src/main/scala/dotty/tools/directives/DirectiveValue.scala:14]: OOP disguised as FP :-)

File: directives-parser/src/main/scala/dotty/tools/directives/Parser.scala [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+import scala.annotation.tailrec
+
+/** Phase 3: recursive-descent parser that consumes a token stream and produces [[UsingDirective]]
+  * nodes.
+  *
+  * Grammar:
+  * {{{
+  * Directives ::= { Directive }
+  * Directive  ::= Using Key Values Newline
```
Comment (SolalPirelli) [inline directives-parser/src/main/scala/dotty/tools/directives/Parser.scala]: this grammar differs from the readme; maybe better to refer to the readme instead?

File: directives-parser/src/main/scala/dotty/tools/directives/Token.scala [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+  /** End of token stream. */
+  case Eof(pos: Position)
+
+  /** A lexer error. */
+  case LexError(message: String, pos: Position)
+
+object Token:
+  extension (t: Token)
+    def pos: Position = t match
```
Comment (SolalPirelli) [inline directives-parser/src/main/scala/dotty/tools/directives/Token.scala:33]: presumably Token could have a `val pos: Position` directly, I guess it doesn't matter much

File: directives-parser/src/test/scala/dotty/tools/directives/CommentExtractorTests.scala [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+    assertEquals(0, r.directiveLines.length)
+    assertEquals(1, r.codeOffset) // BOM(1) + start of `val`
+    val ws = r.diagnostics.filter(_.severity == DiagnosticSeverity.Warning)
+    assertEquals(1, ws.length)
+    assertTrue(ws.head.message.contains("Ignoring"))
+  }
+
+  // -----------------------------------------------------------------------
+  // Edge Case 7: `*/` followed by directive on same line
+  // -----------------------------------------------------------------------
```
Comment (SolalPirelli) [inline directives-parser/src/test/scala/dotty/tools/directives/CommentExtractorTests.scala]: where are the other 6?

**Review by SolalPirelli — APPROVED** [review]
very nice, a few questions

File: directives-parser/README.md (hunk 4 above)
Comment (Gedochao) [reply directives-parser/README.md:52]: ...I suppose it was just implemented this way. I think this could be loosened up as a follow-up.

Comment (Gedochao) [reply directives-parser/README.md:52]: I don't think we want to allow zero spaces for readability's sake, but more whitespace should be allowable (as long as it's not a newline or anything).
But then... I sort of like this being strict, now that I think about it. Maybe I did it deliberately at the time 🤔 As readability increases with it always being a single space IMO.

Comment (Gedochao) [reply directives-parser/README.md:52]: Fun fact: the old parser used to allow multiline directives. 🙃

File: directives-parser/src/main/scala/dotty/tools/directives/DirectiveValue.scala (hunk 7 above)
Comment (Gedochao) [reply directives-parser/src/main/scala/dotty/tools/directives/DirectiveValue.scala:14]: You're leading somewhere here, aren't you 😅

Comment (Gedochao) [reply directives-parser/src/main/scala/dotty/tools/directives/DirectiveValue.scala:14]: I guess that's an area a follow-up refactor would be okay (but I'd avoid further refactorings in this PR)

File: directives-parser/README.md (hunk 4 above)
Comment (SolalPirelli) [reply directives-parser/README.md:52]: I think for consistency with the rest of the language it'd be good to allow any number of spaces including 0, but indeed that's for a future discussion.

File: directives-parser/src/main/scala/dotty/tools/directives/Token.scala (hunk 9 above)
Comment (Gedochao) [reply directives-parser/src/main/scala/dotty/tools/directives/Token.scala:33]: similarly, that's the kind of refactor I'd say is out of scope here

File: directives-parser/src/test/scala/dotty/tools/directives/CommentExtractorTests.scala (hunk 10 above)
Comment (Gedochao) [reply directives-parser/src/test/scala/dotty/tools/directives/CommentExtractorTests.scala]: That may be a leftover comment... lemme clean it up

File: directives-parser/src/main/scala/dotty/tools/directives/Diagnostics.scala (hunk 6 above)
Comment (Gedochao) [reply directives-parser/src/main/scala/dotty/tools/directives/Diagnostics.scala:9]: in Scala CLI some of them don't have positions, as they may be coming from the command line, an environment variable and such.
That said, I suppose we could drop the `Option`, since using directives would always have a position.

Comment (Gedochao) [reply directives-parser/src/main/scala/dotty/tools/directives/Diagnostics.scala:9]: anyway, that could be another follow-up

File: directives-parser/src/test/scala/dotty/tools/directives/CommentExtractorTests.scala (hunk 10 above)
Comment (Gedochao) [reply directives-parser/src/test/scala/dotty/tools/directives/CommentExtractorTests.scala]: done

File: directives-parser/src/main/scala/dotty/tools/directives/Parser.scala (hunk 8 above)
Comment (Gedochao) [reply directives-parser/src/main/scala/dotty/tools/directives/Parser.scala]: cleaned it up

File: project/Build.scala (hunk 3 above)
Comment (Gedochao) [reply project/Build.scala]: fixed
